### PR TITLE
BEAC-167: Move object store configs to object_store_factory

### PR DESCRIPTION
**Pull request by lizardoluis**
This PR moves the object store configuration structs into the new crate `object_store_factory`.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,8 @@ futures = "0.3"
 
 itertools = ">=0.10.0"
 object_store = { version = "0.10.2", features = ["aws", "azure", "gcp"] }
+
+serde = "1.0.156"
```
Comment (gruuya) [inline Cargo.toml:26]: We should use this in `[dependencies]` below.

File: src/config/schema.rs [hunk 2]
```diff
@@ -52,30 +53,6 @@ pub struct SeafowlConfig {
...
 
-#[derive(Deserialize, Debug, PartialEq, Eq, Clone)]
-#[serde(tag = "type", rename_all = "snake_case")]
-pub enum ObjectStore {
-    Local(Local),
-    #[serde(rename = "memory")]
-    InMemory(InMemory),
-    #[cfg(feature = "object-store-s3")]
-    S3(S3),
-    #[cfg(feature = "object-store-gcs")]
```
Comment (gruuya) [inline src/config/schema.rs:63]: We should probably get rid of these two features in Cargo.toml (not used in the new configs, and I don't think we ever opt-out of these).

File: src/config/schema.rs [hunk 3]
```diff
@@ -923,6 +786,7 @@ cache_control = "private, max-age=86400"
     fn test_parse_config_invalid_s3() {
         let error =
             load_config_from_string(TEST_CONFIG_INVALID_S3, false, None).unwrap_err();
+        println!("##### ERROR: {:?}", error.to_string());
```
Comment (gruuya) [inline src/config/schema.rs:789]:
```suggestion
```

File: src/object_store/wrapped.rs [hunk 4]
```diff
@@ -89,16 +85,15 @@ impl InternalObjectStore {
...
+        let prefix = match self.config.clone() {
+            ObjectStoreConfig::AmazonS3(aws_config) => aws_config.prefix,
+            ObjectStoreConfig::GoogleCloudStorage(google_config) => google_config.prefix,
+            _ => None,
+        };
+
+        match prefix {
+            Some(prefix) => Path::from(format!("{prefix}/{table_prefix}")),
+            None => Path::from(table_prefix),
         }
```
Comment (gruuya) [inline src/object_store/wrapped.rs:97]: Seems simpler to do it with one pattern match as before?

```suggestion
        match self.config.clone() {
            ObjectStoreConfig::AmazonS3(S3Config {
                prefix: Some(prefix),
                ..
            })
            | ObjectStoreConfig::GoogleCloudStorage(GCSConfig {
                prefix: Some(prefix),
                ..
            }) => Path::from(format!("{prefix}/{table_prefix}")),
            _ => Path::from(table_prefix),
        }
```

**Review by gruuya — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (lizardoluis) [reply Cargo.toml:26]: @gruuya I get an error if I do not set it here, as I'm inheriting it in the `object_store_factory` Cargo.toml. Btw, it is also set in the `[dependecies]`.

The error is: 

```
error: failed to load manifest for workspace member `/Users/lizardo/Workspace/seafowl/object_store_factory`
referenced by workspace at `/Users/lizardo/Workspace/seafowl/Cargo.toml`

Caused by:
  failed to parse manifest at `/Users/lizardo/Workspace/seafowl/object_store_factory/Cargo.toml`

Caused by:
  error inheriting `serde` from workspace root manifest's `workspace.dependencies.serde`

Caused by:
  `dependency.serde` was not found in `workspace.dependencies
```

File: src/config/schema.rs (hunk 3 above)
Comment (lizardoluis) [reply src/config/schema.rs:789]: Missed that

File: src/object_store/wrapped.rs (hunk 4 above)
Comment (lizardoluis) [reply src/object_store/wrapped.rs:97]: I reverted, though I think the readability was better with the change.

File: Cargo.toml (hunk 1 above)
Comment (gruuya) [reply Cargo.toml:26]: No, i meant for serde to remain in `workspace.dependencies`, just to change line 129/131 in this file to inherit from it `serde = { workspace = true }`.

File: src/config/schema.rs (hunk 2 above)
Comment (lizardoluis) [reply src/config/schema.rs:63]: I purged also some occurrences of it in the code.

File: Cargo.toml (hunk 1 above)
Comment (lizardoluis) [reply Cargo.toml:26]: Ahh, got it.

**Pull request merged**